### DISCUSSION TITLE
Unity plug-in

### DISCUSSION THREAD

**Discussion by erez-tom** [google-deepmind/mujoco#125] [Announcements]
Today we released a MuJoCo plug-in for the [Unity game engine](https://unity.com).  For more details see:
- The associated [documentation chapter](https://mujoco.readthedocs.io/en/latest/unity.html).
- [C# Source code](https://github.com/deepmind/mujoco/tree/main/unity).
- A [Screencast](https://www.youtube.com/watch?v=eyzzsGJ1iic) of the plug-in in action.

**Comment by SanggunLee** [comment]
Hi. Thanks for your work.
Do you have a plan to support UnrealEngine ?

Unity is easy to use and ML engineers (I think mujoco is also used by ML engineers) can use easily.

But, If project is big and your project needs high performence (simulation needs very high performance) , Unreal Engine is suitable.

If users need only ML trading tool, Unity is OK.
If users want to make professional solution, Unreal Engine is needed.
If users dont want coding and just want to use tools, Unity is OK.

If DeepMind has no plan to use UnrealEngin,
I plan to develop UnrealEngine Plugin.
Unity has obvious limitations.

As a result,
Unity is suitable for individuals or education.
UE is suitable for companies and professional use.

**Comment by saran-t** [comment]
@SanggunLee We don't have any plan to develop a plugin for Unreal. If you're interested in doing so yourself please feel free to go ahead. Thanks for your interest!

**Comment by SanggunLee** [comment]
@saran-t  Thanks for reply. I hope I can contribute.This will result in a lot of fun things.

**Comment by JeanElsner** [comment]
Hey guys, the plug-in is looking great so far, very well done.
Do you have any plans to support import using the binary format as well?
That way you could pack a whole scene including assets into a single file, without having to worry about the mesh paths etc.
Similarly it would be very useful if MuJoCo were to support more (versatile) mesh formats or users could export a scene from Unity.

You may have guessed it :-) but I'm very interested in having the simulation and visualization in Unity decoupled (or even distributed). My early tests show that the FixedUpdate method is not completely independent from the framerate and the simulation frequency will ultimately suffer (even if there is enough CPU time to e.g. run a separate MuJoCo instance). As we are using MuJoCo with hardware-in-the-loop in my lab, we need to run both at a high and reliable frequency. Currently I use the native viewer remotely to render the simulation but if possible I'd like to upgrade to something more fancy without interrupting our current workflow too much.
